core/procedural-primitives/procedural-primitives.zig

```diff
@@ -0,0 +1,295 @@
+const std = @import("std");
+const zmath = @import("zmath");
+
+const PI = 3.1415927410125732421875;
+
+pub const F32x3 = @Vector(3, f32);
+pub const F32x4 = @Vector(4, f32);
+pub const VertexData = struct {
+    position : F32x3,
+    normal : F32x3,
+};
+
+pub const PrimitiveType = enum(u4) {
+    none,
+    triangle,
+    quad,
+    plane,
+    circle,
+    uv_sphere,
+    ico_sphere,
+    cylinder,
+    cone,
+    torus
+};
+
+pub const Primitive = struct {
+    vertex_data : std.ArrayList(VertexData),
+    vertex_count : u32,
+    index_data : std.ArrayList(u32),
```

Comment:
Instead of having each primitive generation function allocate new ArrayList's, it might be worth passing those into the generation function. That way the user could reuse memory. especially since it may be short-lived anyway (e.g. until you upload to the GPU.)

If generators could also calculate how much space they will need, it would also let the user allocate one big buffer and then ask the generators to append their results - which would be more efficient if you have a ton of tiny shapes.

Reply:
Thanks! This is a really good idea! I agree with the point of allocating memory in advance! Will do it with next PR